src/ipgroups.cpp

```diff
@@ -308,3 +314,88 @@ void InitIPGroups(CScheduler *scheduler) {
     //   short lived connections and roam around different IPs, but we still want to serve them ahead of long term
     //   idling connections.
 }
+
+IPGroupSlot::IPGroupSlot(const std::string& groupName) :
+    groupName(groupName), groupCS(cs_groups)
+{
+    LOCK(*cs_groups);
+
+    typedef std::vector<CIPGroup>::iterator auto_;
+    for (auto_ g = groups.begin(); g != groups.end(); ++g) {
+        if (g->header.name != groupName)
+            continue;
+
+        g->header.connCount += 1;
+
+        if (g->header.decrPriority)
+            g->header.priority -= IPGROUP_CONN_MODIFIER;
+
+        return; // names are unique, so we're done.
+    }
+    assert(!"tried to create a groupslot for a group that does not exist");
+}
+
+IPGroupSlot::~IPGroupSlot() {
+    // groupCS can be null when exiting bitcoind due to CNetCleanup
+    // instance in net.cpp destructing after cs_groups.
+    boost::shared_ptr<CCriticalSection> cs(groupCS.lock());
+    if (cs.get() == NULL)
+        return;
+
+    LOCK(*cs);
+
+    typedef std::vector<CIPGroup>::iterator auto_;
+    for (auto_ g = groups.begin(); g != groups.end(); ++g) {
+        CIPGroupData& group = g->header;
+
+        if (group.name != groupName)
+            continue;
+
+        group.connCount -= 1;
+        assert(group.connCount >= 0);
+
+        if (group.decrPriority)
+            group.priority += IPGROUP_CONN_MODIFIER;
+
+        if (group.selfErases && !group.connCount)
+            MaybeRemoveGroup(g->header.name);
+
+        return;
+    }
+}
+
+CIPGroupData IPGroupSlot::Group() {
+    LOCK(*cs_groups);
+    typedef std::vector<CIPGroup>::iterator auto_;
+    for (auto_ g = groups.begin(); g != groups.end(); ++g)
+        if (g->header.name == groupName)
+            return g->header;
+
+    return CIPGroupData();
+}
+
+std::auto_ptr<IPGroupSlot> AssignIPGroupSlot(const CNetAddr& ip) {
+    LOCK(*cs_groups);
+
+    CIPGroupData group = FindGroupForIP(ip);
+
+    // IP belongs to a group already.
+    if (!group.name.empty())
+        return std::auto_ptr<IPGroupSlot>(new IPGroupSlot(group.name));
+
+    // If IP does not belong to a group, create a new group
+    // for only this IP. This group is used to de-prioritize multiple
+    // connections from same IP.
+    //
+    // The goal is to force an attacker to spread out in order to get lots of priority.
+
+    // IPGROUP_CONN_MODIFIER will be decremented again when slot is constructed.
+    int pri = DEFAULT_IPGROUP_PRI + IPGROUP_CONN_MODIFIER;
+
+    CIPGroup newGroup;
+    newGroup.header = CIPGroupData(ip.ToStringIP(), pri, true, true);
+    newGroup.subnets.push_back(CSubNet(ip.ToStringIP() + "/32"));
```

Comment:
Here we could lower the priority per connection for the local subnet instead of just the 1 IP. I'm don't know if that's a good idea or not.